crates/core/src/client.rs

```diff
@@ -290,3 +293,37 @@ async fn sn_state_block_number_inner(
 
     Ok(U256::from_big_endian(&sn_block_num).as_u64())
 }
+
+async fn wrapped_retry<'a, Fut, T, E, D>(
+    func: impl Fn(&'a Client<D>, Address) -> Fut,
+    l1_client: &'a Client<D>,
+    core_contract_addr: Address,
+) -> T
+where
+    Fut: Future<Output = Result<T, E>>,
+    D: Database,
+    E: Display,
+{
+    // configure parameters
+    let timeout_secs: u64 = 60;
+    let retry_limit: u8 = 5;
+    let retry_secs: u64 = 5;
+
+    let mut retries: u8 = 0;
+    loop {
+        if retries >= retry_limit {
+            panic!("too many retries");
+        }
+        retries += 1;
+        match future::timeout(time::Duration::from_secs(timeout_secs), func(l1_client, core_contract_addr)).await {
+            Ok(Ok(ret)) => break ret,
+            Ok(Err(err)) => {
+                warn! {"Error for {} times: {}", retries, err}
+            }
+            Err(err) => {
+                warn! {"Timeout for {} times: {}", retries, err}
+            }
+        }
+        thread::sleep(time::Duration::from_secs(retry_secs));
```

Comment:
I'm not sure I understand, it seems we don't need to await `thread::sleep`?

Reply:
`std::thread::sleep` is synchronous and will keep the future runtime worker for as long as needed. This may lead to worker exhaustion, preventing other futures from being executed concurrently.
You'll probably want `async_std::task::sleep` here, this way the runtime worker will be able to pick up some other work while this future is sleeping.

Reply:
Oh, got it! Thanks for helping, this is now fixed ;)